src/ape_cache/query.py

```diff
@@ -0,0 +1,312 @@
+from ape.utils import singledispatchmethod  # type: ignore
+
+from . import models
+from .models import Blocks, ContractEvents, Transactions
+
+
+class CacheQueryProvider(QueryAPI):
+    """
+    Default implementation of the ape.api.query.QueryAPI
+    Allows for the query of blockchain data using connected provider
+    """
+
+    # Database management
+    def get_database_file(self, ecosystem_name: str, network_name: str) -> Path:
+        """
+        Let's us figured out what the file *will be*, mostly used for database management
+        """
+
+        if network_name == LOCAL_NETWORK_NAME:
+            # NOTE: no need to cache local network, no use for data
+            raise QueryEngineError("Cannot cache local data")
+
+        if "-fork" in network_name:
+            # NOTE: send query to pull from upstream
+            network_name = network_name.replace("-fork", "")
+
+        return self.config_manager.DATA_FOLDER / ecosystem_name / network_name / "cache.db"
+
+    def get_sqlite_uri(self, database_file: Path) -> str:
+        return f"sqlite:///{database_file}"
+
+    def init_database(self, ecosystem_name: str, network_name: str):
+        database_file = self.get_database_file(ecosystem_name, network_name)
+        if database_file.is_file():
+            raise QueryEngineError("Database has already been initialized")
+
+        # NOTE: Make sure database folder location has been created
+        database_file.parent.parent.mkdir(exist_ok=True)
+        database_file.parent.mkdir(exist_ok=True)
+
+        models.Base.metadata.create_all(  # type: ignore
+            bind=create_engine(self.get_sqlite_uri(database_file), pool_pre_ping=True)
+        )
+
+    def purge_database(self, ecosystem_name: str, network_name: str):
+        database_file = self.get_database_file(ecosystem_name, network_name)
+        if not database_file.is_file():
+            raise QueryEngineError("Database must be initialized")
+
+        database_file.unlink()
+
+    # Operations
+    @property
+    def database_connection(self):
+        """
+        Gets the currently active network's database, for actual usage
+        NOTE: Makes the database if it doesn't actually exist
+        """
+
+        if not self.network_manager.active_provider:
+            raise QueryEngineError("Not connected to a provider")
+
+        ecosystem_name = self.provider.network.ecosystem.name
+        network_name = self.provider.network.name
+
+        database_file = self.get_database_file(ecosystem_name, network_name)
+        if not database_file.is_file():
+            raise QueryEngineError("Database has not been initialized")
+
+        try:
+            sqlite_uri = self.get_sqlite_uri(database_file)
+            return create_engine(sqlite_uri, pool_pre_ping=True).connect()
+
+        except QueryEngineError as e:
+            logger.debug(f"Exception when querying:\n{e}")
+            return None
+
+        # TODO: Are there any other errors to handle?
```

Comment:
TODO is not really needed, we will handle errors as they come up or else we won't.